### PR TITLE
Move libnetwork

### PR DESCRIPTION
**Pull request by cpuguy83**
This moves libnetwork in-tree as discussed in https://github.com/moby/libnetwork/issues/2522

### CODE REVIEW

**Comment by cpuguy83** [comment]
Bunch of commits w/o a DCO...

**Comment by olljanat** [comment]
afaiu it should be possible to re-trigger CI from Jenkins console with `dco = false` selection so other tests would run normally.:
https://github.com/moby/moby/blob/bedefacfd5cfe332d722439aacd7d3ef4f1e723c/Jenkinsfile#L21

File: Jenkinsfile
Comment (olljanat) [inline Jenkinsfile:21]: Unfortunately Jenkins uses Jenkinsfile from master branch so this does not help...

Comment (thaJeztah) [reply Jenkinsfile]: @cpuguy83 you can use "build with parameters" instead

Comment (cpuguy83) [reply Jenkinsfile]: I don't see a way to. Funny thing is last time it totally built and didn't run the dco check.

Comment (thaJeztah) [reply Jenkinsfile]: Yeah, Jenkins often caches parameters; sometimes it keeps the old ones; it's in the "classic" view, then click the PR number, and then it shows in the left-hand menu 
8461F156-B950-4B02-8106-6C54F2A643C8
CB351E73-B2AE-4519-9499-4755BD2E524B

Comment (cpuguy83) [reply Jenkinsfile]: Oh right! Classic view.

**Comment by thaJeztah** [comment]
I manually marked the DCO _bot_ to "pass" (it has a button for that)

**Review by tianon — COMMENTED** [review]
Generally +1, although the necessary swarmkit and buildkit changes are going to be interesting to resolve. :grimacing:

File: libnetwork/Makefile
Comment (cpuguy83) [reply libnetwork/Makefile:1]: No phony's allowed?

Comment (tianon) [reply libnetwork/Makefile:1]: Oh no, that's fine -- I meant the whole `Makefile`.

I guess what I'm getting at is that it seems odd to bring libnetwork into the moby/moby repo but still treat it as a separate project with its own build scripts, documentation, etc?

**Comment by cpuguy83** [comment]
Rebased and fixed issues:
1. Using proper filter-branch (rather filter-repo) to move history
2. Accidental removal of vendor/archive
3. Newlines for new scripts

re: project stuff underneath the `libnetwork` dir
It would be nice to figure out what to do with these as a follow-up since working out all these changes is not completely trivial, especially with merge conflicts in vendor.conf
We'll also need to setup libnetwork integration tests with CI.

**Comment by cpuguy83** [comment]
@stefanberger Any idea why CI would be complaining about a missing module on the Windows machines for these libnetwork tests?

**Comment by cpuguy83** [comment]
Sorry... wrong Stefan...

**Comment by thaJeztah** [comment]
/cc @kevpar @daschot perhaps you have an idea on this error;

```
=== RUN   TestNAT
--- FAIL: TestNAT (0.00s)
    windows_test.go:34: Failed to create bridge: failed during hnsCallRawResponse: hnsCall failed in Win32: The specified module could not be found. (0x7e)
```

**Comment by cpuguy83** [comment]
@StefanScherer Any idea why CI would be complaining about a missing module on the Windows machines for these libnetwork tests?

**Comment by StefanScherer** [comment]
@cpuguy83 I have no idea why the `hnsCall failed in Win32: The specified module could not be found. (0x7e)` error appears. The Windows agents are Windows Server 2019 from mid February with Docker 19.03.12 installed. I'm planning to update the agents to have all latest updates and Docker 20.10.6 installed.

File: Jenkinsfile
Comment (thaJeztah) [inline Jenkinsfile]: ? whoops?

File: Jenkinsfile
Comment (thaJeztah) [inline Jenkinsfile]: Ah! Hm.. if you need to do another fix, perhaps you could fix the first commit

File: libnetwork/drivers/windows/windows_test.go
Comment (thaJeztah) [inline libnetwork/drivers/windows/windows_test.go:57]: Perhaps we should create tracking issues for these

File: libnetwork/resolver_test.go
Comment (thaJeztah) [inline libnetwork/resolver_test.go]: Same for this one; we should create a tracking ticket

File: libcontainerd/remote/client.go
Comment (thaJeztah) [inline libcontainerd/remote/client.go:572]: I think this one may be legit, and should probably be:

```suggestion
	for _, m := range index.Manifests {
		m := m
		if m.MediaType == images.MediaTypeContainerd1Checkpoint {
			cpDesc = &m
```

File: libnetwork/drivers/overlay/encryption.go
Comment (thaJeztah) [inline libnetwork/drivers/overlay/encryption.go:632]: Same here; this may be legit; should we capture `sp` inside the loop here? (`sp := sp`)

File: libnetwork/drivers/overlay/encryption.go
Comment (thaJeztah) [inline libnetwork/drivers/overlay/encryption.go:641]: Same here; this may be legit; we probably need to capture `sa` inside the loop (`sa := sa`)

File: libnetwork/drivers/bridge/setup_verify.go
Comment (thaJeztah) [inline libnetwork/drivers/bridge/setup_verify.go:43]: Same here; this may be legit; probably need to capture `addrv6` inside the loop (`addrv6 := addrv6`)

File: libnetwork/diagnostic/server.go
Comment (thaJeztah) [inline libnetwork/diagnostic/server.go:111]: Note: `nolint` comments should not have a leading whitespace (comments for automation don't have a leading whitespace)

Probably slightly more cleaner to use `_` to ignore the error than the `//nolint` comments

File: libnetwork/drivers/bridge/bridge.go
Comment (thaJeztah) [inline libnetwork/drivers/bridge/bridge.go:401]: hehe, I recalled the lib network code when I enabled the `dogsled` linter (best name ever); https://github.com/alexkohler/dogsled

**Comment by cpuguy83** [comment]
This is all green

**Comment by cpuguy83** [comment]
Minus DCO because some of these commits we are pulling in do not have DCO sign-offs.

File: libnetwork/endpoint.go
Comment (thaJeztah) [inline libnetwork/endpoint.go:123]: Probably should do a tracking issue for this if important

File: libnetwork/networkdb/cluster.go
Comment (thaJeztah) [inline libnetwork/networkdb/cluster.go:247]: Do you recall what it was complaining about here (i suspect `rnd.Int63()` not being random?) If you do,  would be great to leave it as a comment. You can use this format:

```suggestion
	randStagger := time.Duration(uint64(rnd.Int63()) % uint64(stagger)) //nolint:gosec // some comment here what we're ignoring
```

File: libnetwork/osl/interface_linux.go
Comment (thaJeztah) [inline libnetwork/osl/interface_linux.go:198]: These were not by accident to support arm32 ?

File: hack/test/unit
Comment (tianon) [inline hack/test/unit:24]:
```suggestion
if grep <<<"${pkg_list}" --quiet --fixed-strings 'libnetwork/drivers/bridge' && ! command -v docker-proxy > /dev/null; then
	hack/make.sh binary-proxy install-proxy
fi
```

**Review by thaJeztah — DISMISSED** [review]

File: libnetwork/CHANGELOG.md
Comment (tianon) [reply libnetwork/CHANGELOG.md:3]: :eyes::question:

File: libnetwork/MAINTAINERS
Comment (tianon) [reply libnetwork/MAINTAINERS:1]: :eyes::question:

Thoughts on whether it still makes sense to keep this file in this repository?

File: vendor.conf
Comment (tianon) [inline vendor.conf:36]: :grin:

File: vendor.conf
Comment (tianon) [inline vendor.conf:149]: :cowboy_hat_face:

File: libnetwork/networkdb/cluster.go
Comment (cpuguy83) [reply libnetwork/networkdb/cluster.go:247]: Ah was trying to add extra comments, didn't realize a 2nd `//` was needed.
This is do to rand use.

File: vendor.conf
Comment (tianon) [inline vendor.conf:65]: :grimacing: fun times

(How many different flag parsers are involved in `moby/moby` post-merge here? :weary:)

**Review by tianon — COMMENTED** [review]
Generally still +1 :smile:

File: libnetwork/osl/interface_linux.go
Comment (cpuguy83) [reply libnetwork/osl/interface_linux.go:198]: They are already unit64, there's no conversion.

File: libnetwork/resolver_test.go
Comment (cpuguy83) [reply libnetwork/resolver_test.go]: I resolved this later on... guess I missed the commit this was in.

File: vendor.conf
Comment (cpuguy83) [reply vendor.conf:149]: Yeah, we'll need to update these projects once merged.

File: libcontainerd/remote/client.go
Comment (samuelkarp) [reply libcontainerd/remote/client.go:572]: The `break` on the next line makes me think this is irrelevant.  The warning is that `m` will have the same address through every iteration of the loop and taking the address could mean reading a different value than you intend; however, since we're immediately ending the loop and no new data will be written into `m` we have the correct value in `cpDesc`.

The `m := m` suggestion should silence the warning and be resilient to further changes in this function so it might be slightly better from a maintenance perspective, but I'm not really concerned with the code as-is.

File: vendor.conf
Comment (thaJeztah) [reply vendor.conf:65]: Yeah, we should move this to cobra. I recall when we moved moby to use Cobra, we created a copy of the `mflags` package to libnetwork.

**Comment by thaJeztah** [comment]
DCO is failing because committer name doesn't match sign-off; wondering if adding an entry to `.mailmap` works for that?

```
Commit sha: 9ced389, Author: akim01, Committer: akim01; Expected "akim01 akim01@fortinet.com", but got "Andrew Kim taeyeonkim90@gmail.com".
```

(we can also mark it to "pass" manually; mostly curious if it takes `.mailmap` into account)

**Comment by cpuguy83** [comment]
There's some missing sign-offs as well.

File: libnetwork/cmd/ovrouter/ovrouter.go
Comment (samuelkarp) [inline libnetwork/cmd/ovrouter/ovrouter.go:166]: `os.Kill` was removed here since it can't be caught anyway?

**Review by samuelkarp — COMMENTED** [review]
I've gotten about a quarter of the way through this so far.

For anyone else looking to review: I found it helpful to check out this PR and the tip of libnetwork and use `git diff` to compare them.  For example: `git diff libnetwork/config/config.go docker/libnetwork/config/config.go` will show me that just the imports have changed even though GitHub's diff view is confused.

File: libnetwork/.circleci/config.yml
Comment (cpuguy83) [reply libnetwork/.circleci/config.yml:1]: No this can be removed. All testing is being done by the jenkins runner.

File: libnetwork/api/api.go
Comment (cpuguy83) [reply libnetwork/api/api.go:1]: If it was removed it was because it came up on the linter that it wasn't used.

File: libnetwork/cmd/ovrouter/ovrouter.go
Comment (cpuguy83) [reply libnetwork/cmd/ovrouter/ovrouter.go:166]: Yep. Linter complained.

File: libnetwork/client/client_service_test.go
Comment (cpuguy83) [reply libnetwork/client/client_service_test.go:91]: On this todo, no I don't think its worth it.
I definitely did not scan the codebase for todo's, though.

Comment (samuelkarp) [reply libnetwork/client/client_service_test.go:91]: Thanks.

I wrote this question before I realized that the content of that file was almost identical; I had thought you added it.  I don't think we need to scan the codebase for existing TODOs.

**Review by AkihiroSuda — APPROVED** [review]
LGTM, let's merge this ASAP to avoid rebase hell

**Review by tianon — APPROVED** [review]

**Comment by cpuguy83** [comment]
Merged per @thaJeztah and @tianon

Will follow-up to address PR comments.

**Comment by AkihiroSuda** [comment]
Can we cherry-pick this into 20.10, and archive the old libnetwork repo?

**Pull request merged**
